jerry-core/debugger/jerry-debugger.c

```diff
@@ -0,0 +1,446 @@
+                                          JERRY_CONTEXT (debugger_byte_code_free_head));
+
+  while (byte_code_free_p != NULL)
+  {
+    jerry_debugger_byte_code_free_t *next_byte_code_free_p;
+    next_byte_code_free_p = JMEM_CP_GET_POINTER (jerry_debugger_byte_code_free_t,
+                                                 byte_code_free_p->next_cp);
+
+    jmem_heap_free_block (byte_code_free_p,
+                          ((size_t) byte_code_free_p->size) << JMEM_ALIGNMENT_LOG);
+
+    byte_code_free_p = next_byte_code_free_p;
+  }
+} /* jerry_debugger_free_unreferenced_byte_code */
+
+/**
+ * Send backtrace.
+ */
+static void
+jerry_debugger_send_backtrace (uint8_t *recv_buffer_p) /**< pointer the the received data */
+{
+  JERRY_DEBUGGER_RECEIVE_BUFFER_AS (jerry_debugger_receive_get_backtrace_t, get_backtrace_p);
+
+  uint32_t max_depth;
+  memcpy (&max_depth, get_backtrace_p->max_depth, sizeof (uint32_t));
+
+  if (max_depth == 0)
+  {
+    max_depth = UINT32_MAX;
+  }
+
+  JERRY_DEBUGGER_SEND_BUFFER_AS (jerry_debugger_send_backtrace_t, backtrace_p);
+
+  JERRY_DEBUGGER_INIT_SEND_MESSAGE (backtrace_p);
+  JERRY_DEBUGGER_SET_SEND_MESSAGE_SIZE_FROM_TYPE (backtrace_p, jerry_debugger_send_backtrace_t);
+  backtrace_p->type = JERRY_DEBUGGER_BACKTRACE;
+
+  vm_frame_ctx_t *frame_ctx_p = JERRY_CONTEXT (vm_top_context_p);
+
+  uint32_t current_frame = 0;
+
+  while (frame_ctx_p != NULL && max_depth > 0)
+  {
+    if (current_frame >= JERRY_DEBUGGER_MAX_SIZE (jerry_debugger_frame_t))
+    {
+      if (!jerry_debugger_send (sizeof (jerry_debugger_send_backtrace_t)))
+      {
+        return;
+      }
+      current_frame = 0;
+    }
+
+    jerry_debugger_frame_t *frame_p = backtrace_p->frames + current_frame;
+
+    jmem_cpointer_t byte_code_cp;
+    JMEM_CP_SET_NON_NULL_POINTER (byte_code_cp, frame_ctx_p->bytecode_header_p);
+    memcpy (frame_p->byte_code_cp, &byte_code_cp, sizeof (jmem_cpointer_t));
+
+    uint32_t offset = (uint32_t) (frame_ctx_p->byte_code_p - (uint8_t *) frame_ctx_p->bytecode_header_p);
+    memcpy (frame_p->offset, &offset, sizeof (uint32_t));
+
+    frame_ctx_p = frame_ctx_p->prev_context_p;
+    current_frame++;
+    max_depth--;
+  }
+
+  size_t message_size = current_frame * sizeof (jerry_debugger_frame_t);
+
+  JERRY_DEBUGGER_SET_SEND_MESSAGE_SIZE (backtrace_p, 1 + message_size);
+  backtrace_p->type = JERRY_DEBUGGER_BACKTRACE_END;
+
+  jerry_debugger_send (sizeof (jerry_debugger_send_type_t) + message_size);
+} /* jerry_debugger_send_backtrace */
+
+/**
+ * Check received packet size.
+ */
+#define JERRY_DEBUGGER_CHECK_PACKET_SIZE(type) \
+  if (message_size != sizeof (type)) \
+  { \
+    jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Invalid message size\n"); \
+    jerry_debugger_close_connection (); \
+    return false; \
+  }
+
+/**
+ * Receive message from the client.
+ *
+ * @return true - if message is processed successfully
+ *         false - otherwise
+ */
+inline bool __attr_always_inline___
```

Comment:
I think this function doesn't need to be inlined, debugging code is not realy a performance-critical task :) (and assuming the compiler actually inlines this function this will increase the code size footprint)